rasa/engine/graph.py

```diff
@@ -0,0 +1,198 @@
+        ...
+
+
+@dataclass
+class ExecutionContext:
+    """Holds information about a single graph run."""
+
+    graph_schema: GraphSchema = field(repr=False)
+    model_id: Text
+    diagnostic_data: bool = False
+    is_finetuning: bool = False
+
+
+class GraphNode:
+    """Instantiates and runs a `GraphComponent` within a graph.
+
+    A `GraphNode` is a wrapper for a `GraphComponent` that allows it to be executed
+    In the context of a graph. It is responsible for instantiating the component at the
+    correct time, collecting the inputs from the parent nodes, running the run function
+    of the component and passing the output onwards.
+    """
+
+    def __init__(
+        self,
+        node_name: Text,
+        component_class: Type[GraphComponent],
+        constructor_name: Text,
+        component_config: Dict[Text, Any],
+        fn_name: Text,
+        inputs: Dict[Text, Text],
+        eager: bool,
+        execution_context: ExecutionContext,
+    ) -> None:
+        """Initializes `GraphNode`.
+
+        Args:
+            node_name: The name of the node in the schema.
+            component_class: The class to be instantiated and run.
+            constructor_name: The method used to instantiate the component.
+            component_config: Config to be passed to the component.
+            fn_name: The function to be run when the node executes.
+            inputs: A map from input name to parent node name that provides it.
+            eager: Determines if the node is instantiated right away, or just before
+                being run.
+            execution_context: Information about the current graph run.
+        """
+        self._node_name: Text = node_name
+        self._component_class: Type[GraphComponent] = component_class
+        self._constructor_name: Text = constructor_name
+        self._constructor_fn: Callable = getattr(
+            self._component_class, self._constructor_name
+        )
+        self._component_config: Dict[Text, Any] = {
+            **self._component_class.default_config,
+            **component_config,
+        }
+        self._fn_name: Text = fn_name
+        self._fn: Callable = getattr(self._component_class, self._fn_name)
+        self._inputs: Dict[Text, Text] = inputs
+        self._eager: bool = eager
+        self._execution_context: ExecutionContext = execution_context
+
+        self._component: Optional[GraphComponent] = None
+        if self._eager:
+            self._load_component()
+
+    def _load_component(
+        self, additional_kwargs: Optional[Dict[Text, Any]] = None
+    ) -> None:
+        kwargs = additional_kwargs if additional_kwargs else {}
+
+        logger.debug(
+            f"Node {self._node_name} loading "
+            f"{self._component_class.__name__}.{self._constructor_name} "
+            f"with config: {self._component_config}, and kwargs: {kwargs}."
+        )
+        self._component: GraphComponent = getattr(  # type: ignore[no-redef]
```

Comment:
* should we catch the case when the component for some reason doesn't have the method
* Do we need to handle the case where something goes wrong during the instantiation of a component?

Reply:
> should we catch the case when the component for some reason doesn't have the method

I didn't do this because it will be checked by the validation. However we could have another check here for added safety.

> Do we need to handle the case where something goes wrong during the instantiation of a component?

Wouldn't this require a very broad try catch?

Reply:
> Wouldn't this require a very broad try catch?
 
yeah 🤷🏻  You mean we should just let it raise and fail then?